### PR TITLE
feat(packageRules): matchJsonata

### LINKED ISSUE #31820: packageRules.matchJsonata
### Describe the proposed change(s).

Add new config option `matchJsonata` with parent `packageRules`, to sit alongside `matchPackageNames`, `matchCategories`, etc.

`matchJsonata` should accept a string.

When used, it should use JSONata (npm package `jsonata`) to query the string against the full config object which it is passed, and return `true` or `false`.

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Add `matchJsonata` configuration option to `packageRules`.

## Context

Closes #31820

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/validation.ts [hunk 1]
```diff
@@ -442,6 +442,7 @@ export async function validateConfig(
               'matchCurrentAge',
               'matchRepositories',
               'matchNewValue',
+              'matchJsonata',
```
Comment (astellingwerf) [inline lib/config/validation.ts:445]: Would it make sense to change this hardcoded array into something like?:
```
getOptions().filter(o => o.parents?.includes('packageRules') && o.name.startsWith('match')).map(o => o.name);
```

Comment (rarkins) [reply lib/config/validation.ts:445]: Yeah, we could do that now

Comment (viceice) [reply lib/config/validation.ts:445]: but out of scope for this pr

File: lib/util/package-rules/base.ts [hunk 2]
```diff
@@ -11,5 +11,5 @@ export abstract class Matcher implements MatcherApi {
   abstract matches(
     inputConfig: PackageRuleInputConfig,
     packageRule: PackageRule,
-  ): boolean | null;
+  ): boolean | null | Promise<boolean | null>;
```
Comment (viceice) [inline lib/util/package-rules/base.ts:14]: unrelated change

**Review by viceice — COMMENTED** [review]
LGTM besides the unrelated changes

File: lib/util/package-rules/jsonata.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+export class JsonataMatcher extends Matcher {
+  override async matches(
+    inputConfig: PackageRuleInputConfig,
+    { matchJsonata }: PackageRule,
+  ): Promise<boolean | null> {
+    if (!matchJsonata) {
+      return null;
+    }
+
+    let expression;
```
Comment (viceice) [inline lib/util/package-rules/jsonata.ts:15]: add type

File: lib/config/validation.ts (hunk 1 above)
Comment (astellingwerf) [reply lib/config/validation.ts:445]: It would have prevented the validation error on the first attempt at this change set...

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -2856,6 +2856,21 @@ The following example matches any file in directories starting with `app/`:
...
+See [the JSONata docs](https://docs.jsonata.org/) for more details on JSONata syntax.
+
+Here are some example `matchJsonata` strings for inspiration:
+
+```
+$exists(deprecationMessage)
+$exists(vulnerabilityFixVersion)
+manager = 'dockerfile' and depType = 'final'
+```
+
```
Comment (secustor) [inline docs/usage/configuration-options.md:2873]: We should create an example on https://try.jsonata.org/ with the full so people have not to poke around in the dark.

File: lib/config/types.ts [hunk 5]
```diff
@@ -381,6 +381,7 @@ export interface PackageRule
   matchRepositories?: string[];
   matchSourceUrls?: string[];
   matchUpdateTypes?: UpdateType[];
+  matchJsonata?: string;
```
Comment (secustor) [inline lib/config/types.ts:384]: IMO an array would be benefitial here by allowing user to break the logic in separat elements.

Comment (rarkins) [reply lib/config/types.ts:384]: Do you mean instead of using "and"s?

Comment (secustor) [reply lib/config/types.ts:384]: I mean OR chains

Comment (rarkins) [reply lib/config/types.ts:384]: Doesn't this then make it harder to debug eg in the playground?

Maybe we can start this way with single string and see if its really a problem

Comment (secustor) [reply lib/config/types.ts:384]: That way it is possible to test each element independently.  

But fine for me, it would be easy to migrate configs.

File: docs/usage/configuration-options.md (hunk 4 above)
Comment (rarkins) [reply docs/usage/configuration-options.md:2873]: Let's have it as a separate issue though

**Review by viceice — DISMISSED** [review]

File: lib/config/validation.ts [hunk 6]
```diff
@@ -835,6 +837,18 @@ export async function validateConfig(
         }
       }
     }
+
+    if (key === 'matchJsonata' && is.array(val)) {
+      for (const expression of val as string[]) {
```
Comment (secustor) [inline lib/config/validation.ts:842]:
```suggestion
    if (key === 'matchJsonata' && is.array(val, is.string)) {
      for (const expression of val) {
```

**Review by secustor — APPROVED** [review]
